### ISSUE TITLE
VSCode+SSH closes all open tabs on connection failure, losing developer workspace state

### ISSUE THREAD

**Issue by sbrocket** [microsoft/vscode-remote-release#3475]
- VSCode Version: 1.47.3
- Local OS Version: macOS 10.15.6 / Darwin x64 19.6.0
- Remote OS Version: Linux x64 5.2.17-1rodete3-amd64 (internal Debian-based Linux distro)
- Remote Extension/Connection Type: SSH

Steps to Reproduce:

1. Start with an open workspace with one or more open files and SSH connected.
2. SSH connection fails (either manually simulated, e.g. `ssh -O exit <host>` if using multiplexing, or because of other reasons)
    * In my case, "other reasons" happens nightly because the host I'm connecting to involves manual/physical presence authentication steps every 24 hours.
3. Extension attempts to reconnect:
3.a. If it succeeds, everything is fine.
3.b. If it fails, VSCode will pop up a prompt saying "Cannot reconnect. Please reload the window.". This is still OK.
4. Reload the window.
5. Window reloads with all files still open and tries to connect:
5.a. If the connection succeeds, everything is fine and off to the races again.

Expected behavior:

5.b. If the connection fails again, the same or behavior should result; prompt the user to attempt reconnection again. Once connection succeeds, the user's workspace should be the same as before.

Actual behavior:

5.b. If the connection fails again, all open files are closed, workspace split layout is lost, and a different prompt is shown, `Could not establish connection to "X".`
6. This prompt has a retry button. If that's used and the connection succeeds this time, the workspace is not restored.

This issue also occurs if VSCode is closed while in state 3.b. Upon reopening VSCode, if the connection fails 5.b. occurs, same as reloading the window. This scenario often happens to me if I reboot my machine (to apply updates, etc.) while in state 3.b, then VSCode automatically reopens on login before I have a chance to perform the authentication steps that would allow an SSH connection to occur. (This is actually how I encounter this most often.)

I also have a screen recording showing the steps above if that would be useful to attach.

Does this issue occur when you try this locally?: N/A, specific to SSH reconnection
Does this issue occur when you try this locally and all extensions are disabled?: N/A, specific to SSH reconnection

**Comment by alexw23** [comment]
Same issue here. The logic around disconnects needs a rework.

To me the most usable interface would be to simply store the state locally. Upon failure to reconnect with host, server status in bottom right should show "Failed to reconnect". At which point you can click it (the status) to prompt a reconnect. If this fails it would simply go back to the failed status again.

At no point ever should local workspace state be lost.

**Comment by mtaufen** [comment]
Also hitting this problem on a regular basis. This is super annoying and unexpected behavior. Please fix.

**Comment by sbrocket** [comment]
Just got bitten by this again. It's extremely frustrating. Why is this a feature request?

**Comment by alexw23** [comment]
There's a few issues floating around that have either been closed or ignored. Maybe the best option is that we consolidate all "connection" issues into one issue.

**Comment by sbrocket** [comment]
There are other, different reconnection related issues that this shouldn't be grouped with. I intentionally scoped this one to the "losing workspace state" issue. There may be other dupes for this specific thing but I didn't find any before filing initially.

**Comment by MVanderer** [comment]
This issue is the bane of my everyday existence as I work exclusively over ssh. At the very least, could the remote preserve the folders' last used workspace state and make it available to vscode on reconnect?

**Comment by bamurtaugh** [comment]
As an alternate path resulting in lost state that could be related to the same root cause (linked from #3842 so that it doesn't get lost; steps 1&2 from above don't apply as there isn't an open connection and then an SSH connection failure, but just a failure from the get-go):

> If I re-open VSCode having last been working in a remote workspace, but then the log-in fails (say, because I left the prompt open and the server booted me), retrying results in all my open editor tabs being closed, which would not be the case had I successfully logged in the first time around. I don't believe this is the expected behaviour from a user's perspective.
> 
> Extension version: 0.55.0
> VS Code version: Code 1.50.0 (93c2f0fbf16c5a4b10e4d5f89737d9c2c25488a3, 2020-10-07T06:10:52.432Z)
> OS version: Windows_NT x64 10.0.19041
> Remote OS version: Linux x64 4.19.111.1.amd64-smp

**Comment by roblourens** [comment]
I can repro this and the steps explain https://github.com/microsoft/vscode-remote-release/issues/2416 and other issues.

It restores some state like the fact that there was a split, but not which files were open. @bpasero

**Comment by bpasero** [comment]
If we cannot restore an editor on startup because of some connection error, then we close that editor and continue to try to restore the next until we found one that can be restored. We have no concept of showing an error for an editor when it fails to load but keep it open. 

I opened https://github.com/microsoft/vscode/issues/110062 to track that as a feature request. This one can be closed as duplicate.
